nixos/tayga: init

hax404 opened the pull request:
###### Description of changes
Add module and test. Module comes from https://github.com/NixOS/nixpkgs/pull/129390
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
nixos/tayga: init

hax404 committed:
nixos/tests/tayga: init

NickCao commented on nixos/modules/services/networking/tayga.nix:
```diff
@@ -0,0 +1,162 @@
+    data-dir ${cfg.dataDir}
+  '';
+
+  makeRouteCommands = oper: ''
+    ip addr ${oper} ${cfg.ipv4.router.address} dev ${cfg.tunDevice}
+    ip route ${oper} ${strAddr cfg.ipv4.pool} dev ${cfg.tunDevice}
+    ip addr ${oper} ${cfg.ipv6.router.address} dev ${cfg.tunDevice}
+    ip route ${oper} ${strAddr cfg.ipv6.pool} dev ${cfg.tunDevice}
+  '';
```
Could we use `networking` options to implement this instead of ad-hoc iproute2 command?

mweinelt replied:
I already talked to hax404 about this and proposed to use networkd units instead. I'm opposed to relying on the highly undermaintained scripted networking parts of nixos.

hax404 committed:
tayga: correct license to GPL2+

The head of tayga.c says „This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.“

NickCao commented on nixos/modules/services/networking/tayga.nix:
```diff
@@ -0,0 +1,194 @@
+  };
+
+  config = mkIf cfg.enable {
+    systemd.network = {
```
`networking.interfaces` should be used as it is backend agnostic and more future proof.

hax404 commented on nixos/modules/services/networking/tayga.nix:
```diff
@@ -0,0 +1,221 @@
+  };
+
+  config = mkIf cfg.enable {
+    #networking.interfaces."${cfg.tunDevice}" = mkIf (!config.systemd.network.enable) {
```
I would like to check if systemd-networkd is enabled as on other lines. But `mkIf (!config.systemd.network.enable)` runs into infinitie recursion.
```
georg@rick:~/git/nixpkgs/ > nom-build -j0 . -A nixosTests.tayga
/nix/store/23yy920js0pk5924c3899ajf7kxw7i7y-vm-test-run-tayga
Finished at 21:49:58 after 13s
georg@rick:~/git/nixpkgs/ > nom-build -j0 . -A nixosTests.tayga
error: infinite recursion encountered

       at /home/georg/git/nixpkgs/lib/modules.nix:528:12:

          527|         zipAttrsWith (n: concatLists)
          528|           (map (module: let subtree = module.${attr}; in
             |            ^
          529|               if !(builtins.isAttrs subtree) then
(use '--show-trace' to show detailed location information)
Finished at 21:50:06 after 1s
```

NickCao commented:
I'm still confused by the useNetworkd part, why can't we keep only the generic implementation.

hax404 commented:
Now I use the generic implementation. To get the most of the sandboxing, there is a check if systemd-networkd is used, because `virtualOwner` cannot be empty in the nixos-scripts and then uses `root` by defaut.

NickCao commented on nixos/modules/services/networking/tayga.nix:
```diff
@@ -0,0 +1,196 @@
+        ];
+        StateDirectory = "tayga";
+        DynamicUser = mkIf config.networking.useNetworkd true;
+        User = mkIf config.networking.useNetworkd "tayga";
```
User name should default to unit name, no need to set it.

hax404 replied:
thanks

NickCao commented on pkgs/tools/networking/tayga/default.nix:
```diff
@@ -9,6 +9,8 @@ stdenv.mkDerivation rec {
     sha256 = "1700y121lhvpna49bjpssb7jq1abj9qw5wxgjn8gzp6jm4kpj7rb";
   };
 
+  passthru.tests.teeworlds = nixosTests.tayga;
```
Is `teeworlds` is typo?

hax404 replied:
oh yes, thanks

NickCao commented on nixos/tests/all-tests.nix:
```diff
@@ -620,6 +620,7 @@ in {
   systemd-misc = handleTest ./systemd-misc.nix {};
   tandoor-recipes = handleTest ./tandoor-recipes.nix {};
   taskserver = handleTest ./taskserver.nix {};
+  tayga = handleTest ./tayga.nix {};
```
Better to test both the legacy and networkd implementation.

hax404 replied:
Inside the test is done with both the legacy and systemd based network config. That is a small loop in line 251 that tests the routers after each other.

NickCao reviewed (approved):
Apart from the comments, LGTM

RaitoBezarius commented:
Could you move the release notes for 23.05? I am interested into getting this merged.

mweinelt commented:
We're backporting this into 22.11 since it is unlikely to cause regressions.

RaitoBezarius reviewed (approved):
Looks good to me.
